### PR TITLE
Last price animation

### LINKED ISSUE #567: How to add buffer animation to price jump
How to add buffer animation when price jumps
`areaSeries.update`

GIF1

### PR DESCRIPTION
**Pull request by kirchet**
**Type of PR:** enhancement

**PR checklist:**

- [x] Addresses an existing issue: fixes #567
- [x] Includes tests
- [x] Documentation update

**Overview of change:**

Added possibility to enable animation on changing area/line series' last price

### CODE REVIEW

File: src/views/pane/series-last-price-animation-pane-view.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+	public update(updateType?: UpdateType): void {
+		this._invalidated = true;
+		if (updateType === 'data') {
+			if (this._series.options().lastPriceAnimation === LasPriceAnimationMode.OnDataUpdate) {
+				const now = performance.now();
+				const timeToAnimationEnd = this._endTime - now;
+				if (timeToAnimationEnd > 0) {
+					if (timeToAnimationEnd < LasPriceAnimationMode.OnDataUpdate / 4) {
```
Comment (kirchet) [inline src/views/pane/series-last-price-animation-pane-view.ts:143]:
```suggestion
					if (timeToAnimationEnd < Constants.AnimationPeriod / 4) {
```

File: src/views/pane/series-last-price-animation-pane-view.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+	for (const stageData of animationStagesData) {
+		if (globalStage >= stageData.start && globalStage <= stageData.end) {
+			currentStageData = stageData;
+			break;
+		}
+	}
+
+	if (currentStageData === undefined) {
+		throw new Error('Last price animation internal logic error');
```
Comment (timocov) [inline src/views/pane/series-last-price-animation-pane-view.ts:112]: can't we use `ensureDefined`/`assert` functions here?

**Review by timocov — CHANGES_REQUESTED** [review]

File: docs/constants.md [hunk 3]
```diff
@@ -80,3 +80,12 @@ It has the following values:
...
 - `TickMarkType.TimeWithSeconds` - the tick mark represents a full time format.
+
+## LasPriceAnimationMode
+
+`LasPriceAnimationMode` enum is used to specify the type of the last price animation for series such as area or line.
+It has the following values:
+
+- `LasPriceAnimationMode.Disabled` - animation is always disabled.
+- `LasPriceAnimationMode.Continuous` - animation is always enabled.
+- `LasPriceAnimationMode.OnDataUpdate` - animation is active some time after data update.
```
[the remark this answers is not in the document]
Comment (kirchet) [reply docs/constants.md:91]: No, it's not. For now it's 2600 msecs

File: src/views/pane/series-last-price-animation-pane-view.ts (hunk 2 above)
Comment (kirchet) [reply src/views/pane/series-last-price-animation-pane-view.ts:112]: Yes, we can, but if we'll use `assert` we'll need to wrap `currentStageData` with `ensureDefined` as well. If we'll use only `ensureDefined` then it'll be not so informative without this message

Comment (timocov) [reply src/views/pane/series-last-price-animation-pane-view.ts:112]: > but if we'll use assert we'll need to wrap currentStageData with ensureDefined as well

We don't, TypeScript is smart enough to handle it itself: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQwM6oKYCcoAoIIAmMsCAXIgEZxwA2GyYANIgLYbrIDmGA-BaihYYYLgEoKaTDlSICYYqSQBvAFABIGMES4AhPMXwwYxGvXqoACyxwA7ojAZ7AUSw2suAOQBBdNiWIwMgw9ISeiADUOuycPIi8iJ4U4VExqNwYiMmeYmIA3BoAvqrFqoQYELTIWJnygoE0FGAgrJTYiAA+iODlwCIYhAWqwDQAdFBwzgAeAA4IGGCwyLS4+aqqUv64I3CIugC8+90KGH2OhCyeaRk5Qzvjk7PzizDLqwVAA

File: docs/area-series.md [hunk 4]
```diff
@@ -71,6 +72,7 @@ An area series interface can be customized using the following set of options:
         crosshairMarkerRadius: 3,
         crosshairMarkerBorderColor: 'rgb(255, 255, 255, 1)',
         crosshairMarkerBackgroundColor: 'rgb(34, 150, 243, 1)',
+        lastPriceAnimation: 1,
```
[the remark this answers is not in the document]
Comment (timocov) [reply docs/area-series.md:75]: Also, could you enable this new option in coverage tests please?

File: src/model/idata-source.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ export interface IDataSource {
 	priceAxisViews(pane?: Pane, priceScale?: PriceScale): readonly IPriceAxisView[];
 	timeAxisViews(): readonly ITimeAxisView[];
 	paneViews(pane: Pane): readonly IPaneView[];
+	topPaneViews?(pane: Pane): readonly IPaneView[];
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/idata-source.ts:19]: Added docs

File: src/model/series.ts [hunk 6]
```diff
@@ -350,6 +357,20 @@ export class Series<T extends SeriesType = SeriesType> extends PriceDataSource i
...
 	}
 
+	public topPaneViews(pane: Pane): readonly IPaneView[] {
+		const animationPaneView = this._lastPriceAnimationPaneView;
+		if (animationPaneView === null || !animationPaneView.visible()) {
+			return [];
+		}
+
+		if (animationPaneView.animationActive()) {
+			setTimeout(() => this.model().cursorUpdate(), 0);
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/series.ts]: Fixed

File: docs/line-series.md [hunk 7]
```diff
@@ -69,6 +70,7 @@ A line series interface can be customized using the following options:
         crosshairMarkerBorderColor: '#ffffff',
         crosshairMarkerBackgroundColor: '#2296f3',
         lineType: 1,
+        lastPriceAnimation: 1,
```
[the remark this answers is not in the document]
Comment (kirchet) [reply docs/line-series.md]: Fixed

File: docs/area-series.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kirchet) [reply docs/area-series.md]: Done

File: src/views/pane/series-last-price-animation-pane-view.ts (hunk 2 above)
Comment (kirchet) [reply src/views/pane/series-last-price-animation-pane-view.ts]: Fixed

File: src/model/idata-source.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (timocov) [reply src/model/idata-source.ts:19]: Can you point out on them please?

**Review by timocov — APPROVED** [review]

**Pull request merged**